tokio-util/src/codec/any_delimiter_codec.rs

```diff
@@ -0,0 +1,264 @@
+use crate::codec::decoder::Decoder;
+use crate::codec::encoder::Encoder;
+
+use bytes::{Buf, BufMut, BytesMut};
+use std::{cmp, fmt, io, str, usize};
+
+const DEFAULT_DELIMITERS: &str = ",;\n\r";
+const DEFAULT_ENCODERS: &str = ",";
+/// A simple [`Decoder`] and [`Encoder`] implementation that splits up data into chunks based on any character in the given delimiter string.
+///
+/// [`Decoder`]: crate::codec::Decoder
+/// [`Encoder`]: crate::codec::Encoder
+///
+/// # Example
+/// Decode string of bytes containing various different delimiters.
+///
+/// [`BytesMut`]: bytes::BytesMut
+/// [`Error`]: std::io::Error
+///
+/// ```
+/// use tokio_util::codec::{AnyDelimiterCodec, Decoder};
+/// use bytes::{BufMut, BytesMut};
+///
+/// #
+/// # #[tokio::main(flavor = "current_thread")]
+/// # async fn main() -> Result<(), std::io::Error> {
+/// let mut codec = AnyDelimiterCodec::new(",;\r\n",";");
+/// let buf = &mut BytesMut::new();
+/// buf.reserve(200);
+/// buf.put_slice(b"chunk 1,chunk 2;chunk 3\n\r");
+/// assert_eq!("chunk 1", codec.decode(buf).unwrap().unwrap());
+/// assert_eq!("chunk 2", codec.decode(buf).unwrap().unwrap());
+/// assert_eq!("chunk 3", codec.decode(buf).unwrap().unwrap());
+/// assert_eq!("", codec.decode(buf).unwrap().unwrap());
+/// assert_eq!(None, codec.decode(buf).unwrap());
+/// # Ok(())
+/// # }
+/// ```
+///
+#[derive(Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub struct AnyDelimiterCodec {
+    // Stored index of the next index to examine for the delimiter character.
+    // This is used to optimize searching.
+    // For example, if `decode` was called with `abc` and the delimiter is '{}', it would hold `3`,
+    // because that is the next index to examine.
+    // The next time `decode` is called with `abcde}`, the method will
+    // only look at `de}` before returning.
+    next_index: usize,
+
+    /// The maximum length for a given chunk. If `usize::MAX`, chunks will be
+    /// read until a delimiter character is reached.
+    max_length: usize,
+
+    /// Are we currently discarding the remainder of a chunk which was over
+    /// the length limit?
+    is_discarding: bool,
+
+    /// store the delimiter characters in their binary representation
+    delimiters: &'static [u8],
+
+    /// store the encoders characters in their binary representation
+    encoders: &'static [u8],
+}
+
+impl AnyDelimiterCodec {
+    /// Returns a `AnyDelimiterCodec` for splitting up data into chunks.
+    ///
+    /// # Note
+    ///
+    /// The returned `AnyDelimiterCodec` will not have an upper bound on the length
+    /// of a buffered chunk. See the documentation for [`new_with_max_length`]
+    /// for information on why this could be a potential security risk.
+    ///
+    /// [`new_with_max_length`]: crate::codec::AnyDelimiterCodec::new_with_max_length()
+    pub fn new(delimiters: &'static str, encoders: &'static str) -> AnyDelimiterCodec {
```

Comment:
It doesn't seem right to use a `&str` here. If I pass a non-ascii character as delimiter, it will treat any of the bytes that make it up as a delimiter, and prevents me from using non-ascii characters as delimiters.

It also seems limiting to require a `'static` list of delimiters. Then you cannot use the codec with a variable delimiter.

Reply:
I agree with you.
what do you think about changing it from &str to [u8] to find any byte, and this codec use case will be only for simple bytes and not textual meaning?
and another codec that I'm plan to write will treat sequence use case.

Reply:
I'm honestly not sure what a good type would be. Maybe be generic over types that implement the trait `AsRef<[u8]>`?

Reply:
AsRef will be more flexible but I'm afraid that it will confuse with the same issue you point while using &str ?

Reply:
Mm, it's a bit annoying that `str` implements `AsRef<[u8]>` for that, yes.